public/app/features/alerting/unified/components/contact-points/useContactPoints.ts

```diff
@@ -404,100 +392,88 @@ type ContactPointOperationArgs = {
       ? await createOnCallIntegrations(contactPoint)
       : contactPoint;
 
-    if (useK8sApi) {
-      const namespace = getK8sNamespace();
+    const namespace = getK8sNamespace();
+    const contactPointToUse = grafanaContactPointToK8sReceiver(contactPointWithMaybeOnCall);
 
-      const contactPointToUse = grafanaContactPointToK8sReceiver(receiverWithPotentialOnCall);
+    return createGrafanaContactPoint({
+      namespace,
+      comGithubGrafanaGrafanaPkgApisAlertingNotificationsV0Alpha1Receiver: contactPointToUse,
+    }).unwrap();
+  });
 
-      return createGrafanaContactPoint({
-        namespace,
-        comGithubGrafanaGrafanaPkgApisAlertingNotificationsV0Alpha1Receiver: contactPointToUse,
-      }).unwrap();
-    }
+  const updateAlertmanagerConfiguration = useAsync(async ({ contactPoint }: CreateContactPointArgs) => {
+    const contactPointWithMaybeOnCall = isGrafanaAlertmanager
+      ? await createOnCallIntegrations(contactPoint)
+      : contactPoint;
 
-    const config = await getAlertmanagerConfig(alertmanager).unwrap();
-    const newConfig = updateConfigWithReceiver(config, receiverWithPotentialOnCall);
+    const action = addReceiverAction(contactPointWithMaybeOnCall);
+    return produceNewAlertmanagerConfiguration(action);
+  });
 
-    return await dispatch(
-      updateAlertManagerConfigAction({
-        newConfig: newConfig,
-        oldConfig: config,
-        alertManagerSourceName: alertmanager,
-      })
-    )
-      .unwrap()
-      .then(() => {
-        dispatch(alertingApi.util.invalidateTags(['AlertmanagerConfiguration', 'ContactPoint', 'ContactPointsStatus']));
-        dispatch(generatedReceiversApi.util.invalidateTags(['Receiver']));
-      });
-  };
+  return useK8sApi ? updateK8sAPI : updateAlertmanagerConfiguration;
 };
 
+type UpdateContactPointArgsK8s = ContactPointOperationArgs & {
+  /** ID of existing contact point to update - used when updating via k8s API */
+  id: string;
+  resourceVersion?: string;
+};
+type UpdateContactPointArgsConfig = ContactPointOperationArgs & {
+  /** Name of the existing contact point - used for checking uniqueness of name when not using k8s API*/
+  originalName: string;
+};
+type UpdateContactpointArgs = UpdateContactPointArgsK8s | UpdateContactPointArgsConfig;
+
 export const useUpdateContactPoint = ({ alertmanager }: BaseAlertmanagerArgs) => {
   const isGrafanaAlertmanager = alertmanager === GRAFANA_RULES_SOURCE_NAME;
   const useK8sApi = shouldUseK8sApi(alertmanager);
 
   const { createOnCallIntegrations } = useOnCallIntegration();
-  const [getAlertmanagerConfig] = useLazyGetAlertmanagerConfigurationQuery();
   const [replaceGrafanaContactPoint] = useReplaceNamespacedReceiverMutation();
+  const [produceNewAlertmanagerConfiguration] = useProduceNewAlertmanagerConfiguration();
 
-  const dispatch = useDispatch();
+  const updateContactPoint = useAsync(async (args: UpdateContactpointArgs) => {
+    if ('resourceVersion' in args && useK8sApi) {
```

Comment:
I see `resourceVersion` is an optional prop. I'm not sure we should add this expression to decide if we use k8sapi 🤔, because is it possible to have useK8sApi:true and not have `resourceVersion` in the props? if so, then we would use alert manager config api.